src/ray/raylet/runtime_env_agent_client.cc

```diff
@@ -182,6 +189,59 @@ class Session : public std::enable_shared_from_this<Session> {
       RAY_LOG(INFO) << "on_read error after response body received: " << ec.message();
     }
   }
+
+  tcp::resolver resolver_;
+  beast::tcp_stream stream_;
+  std::string host_;
+  std::string port_;
+  std::function<void(std::string)> succ_callback_;
+  std::function<void(ray::Status)> fail_callback_;
+  beast::flat_buffer buffer_;  // (Must persist between reads)
+  http::request<http::string_body> req_;
+  http::response<http::string_body> res_;
+  FinishedCallback finished_callback_;
+};
+
+// A pool of sessions with a fixed max concurrency. Each session can handle 1 concurrent
+// request.
+// Users can post a session to this pool, but should not *run* them. The Session pool runs
+// them if the concurrency is not exceeding the desired value.
+//
+// Once a session is done, remove it from the pool.
+//
+// NOT thread safe: if the methods or the fallbacks are invoked in different threads, the
```

Comment:
Should we add some sort of check in the constructor that checks the main thread id?

Reply:
If the user sticks to a single boost io_context this should be fine. If not we will have to use boost strand to limit it. But today I don't think we have more than 1 io_context anywhere.